rclcpp/include/rclcpp/intra_process_manager.hpp

```diff
@@ -0,0 +1,413 @@
+   * a new intra process message, but it wasn't registered with the manager at
+   * the time of publishing, causing it to take when it wasn't intended.
+   * This should be avioded unless latching-like behavior is involved.
+   *
+   * The message parameter is used to store the taken message.
+   * On the last expected call to this method, the ownership is transfered out
+   * of internal storage and into the message parameter.
+   * On all previous calls a copy of the internally stored message is made and
+   * the ownership of the copy is transfered to the message parameter.
+   * TODO(wjwwood): update this documentation when latching is supported.
+   *
+   * The message parameter can be set to nullptr if:
+   *
+   * - The publisher id is not found.
+   * - The message sequence is not found for the given publisher id.
+   * - The requesting subscription's id is not in the list of intended takers.
+   * - The requesting subscription's id has been used before with this message.
+   *
+   * This method may allocate memory to copy the stored message.
+   *
+   * /param intra_process_publisher_id the id of the message's publisher.
+   * /param message_sequence_number the sequence number of the message.
+   * /param requesting_subscriptions_intra_process_id the subscription's id.
+   * /param message the message typed unique_ptr used to return the message.
+   */
+  template<typename MessageT>
+  void
+  take_intra_process_message(
+    uint64_t intra_process_publisher_id,
+    uint64_t message_sequence_number,
+    uint64_t requesting_subscriptions_intra_process_id,
+    std::unique_ptr<MessageT> & message)
+  {
+    message = nullptr;
+    publisher_info * info;
+    {
+      auto it = publishers_.find(intra_process_publisher_id);
+      if (it == publishers_.end()) {
+        // Publisher is either invalid or no longer exists.
+        return;
+      }
+      info = &it->second;
+    }
+    // Figure out how many subscriptions are left.
+    std::set<uint64_t> * target_subs;
+    {
+      auto it = info->target_subscriptions_by_message_sequence.find(message_sequence_number);
+      if (it == info->target_subscriptions_by_message_sequence.end()) {
+        // Message is no longer being stored by this publisher.
+        return;
+      }
+      target_subs = &it->second;
+    }
+    {
+      auto it = std::find(
+        target_subs->begin(), target_subs->end(),
+        requesting_subscriptions_intra_process_id);
+      if (it == target_subs->end()) {
+        // This publisher id/message seq pair was not intended for this subscription.
+        return;
+      }
+      target_subs->erase(it);
+    }
+    typedef typename mapped_ring_buffer::MappedRingBuffer<MessageT> TypedMRB;
+    typename TypedMRB::SharedPtr typed_buffer = std::static_pointer_cast<TypedMRB>(info->buffer);
+    // Return a copy or the unique_ptr (ownership) depending on how many subscriptions are left.
+    if (target_subs->size()) {
```

Comment:
Another possible race condition:

Suppose there are two target subscribers executing in Thread A and Thread B. If Thread A blocks right after this line and Thread B executes `take_intra_process_message`, Thread B will think it is the last subscriber since Thread A already erased itself from `target_subs`. Thread B will execute `typed_buffer->pop_at_key`, then Thread A will try to get a NULL copy of the message, since it was popped off.